### PR TITLE
Preserve node artefacts for each type of test

### PR DESCRIPTION
**Pull request by jumaffre**
**Before**

Our end-to-end tests deploy "remote" nodes to `/tmp/<user>_<node_id>`, which imply that the logs are deleted between each test run (not useful when findings out why test 22 out of 37 failed on the CI). 

**After**

Nodes are now deployed to a specific `--workspace` directory (defaults to `build/workspace`) and each folder is named after the test name (`--label`) and the node id, e.g. `build/workspace/end_to_end_logging_0`. Note that the executables (cchost, enclave file) are now symlinked instead of being copied to the workspace directories.

-- 
- Declaring an end-to-end test from cmake now requires the `add_e2e_test` wrapper to automatically pass the name of the test as `--label`.
- For `SSHRemote`, nodes are deployed to the same location, e.g. `~/CCF/build/workspace/`. 
- I have also refactored our `quote_verification` end-to-end test so that is now uses our python infra. It required adding new `verify_quote`, `other_quote` and `other_quoted_data` arguments in a few places

### CODE REVIEW

File: tests/memberclient.py [hunk 1]
```diff
@@ -293,13 +293,6 @@ def run(args):
...
-        parser.add_argument(
-            "-p",
-            "--package",
-            help="The enclave package to load (e.g., libsimplebank)",
-            required=True,
-        )
-
-    args = e2e_args.cli_args(add)
+    args = e2e_args.cli_args()
+    args.package = "libloggingenc"
```
Comment (eddyashton) [inline tests/memberclient.py]: We should keep the `--package` arg here, so we can easily run the test with different apps. We can make it non-required, and default to "libloggingenc".

File: sphinx/source/demo.rst [hunk 2]
```diff
@@ -56,5 +56,5 @@ The ``e2e`` test script takes several additional parameters, documented by passi
 
 .. rubric:: Footnotes
 
-.. [#log_location] The log location should be visible in the Python output. By default, node ``N`` will log to files ``out`` and ``err`` in ``/tmp/{USER}_{N}``
+.. [#log_location] The log location should be visible in the Python output. By default, node ``N`` will log to files ``out`` and ``err`` in ``CCF/build/ccf_workspace/end_to_end_scenario_{N}``
```
Comment (achamayou) [inline sphinx/source/demo.rst]: s/ccf_workspace/ws/, we know it's CCF. Perhaps test_ws to make it clear it's tests.

File: tests/client.py [hunk 3]
```diff
@@ -54,7 +54,9 @@ def cli_args(add=lambda x: None, accept_unknown=False):
     )
 
     # Some common client binary args are exposed here for ease
-    parser.add_argument("--label", help="Test label name in perf summary output file")
+    parser.add_argument(
```
Comment (achamayou) [inline tests/client.py]: This should default to something like basename(argv[1]) with the suffix stripped I think. And it should maybe just be a suffix for when we run the same thing multiple times with variants of the arguments.

File: tests/e2e_args.py [hunk 4]
```diff
@@ -82,6 +83,12 @@ def cli_args(add=lambda x: None, parser=None, accept_unknown=False):
     parser.add_argument(
         "--notify-server", help="Server host to notify progress to (host:port)"
     )
+    parser.add_argument(
+        "--workspace",
+        help="Directory to setup nodes",
```
Comment (achamayou) [inline tests/e2e_args.py]: "Temporary directory where nodes store their logs, ledgers, quotes etc."

File: tests/e2e_args.py [hunk 5]
```diff
@@ -82,6 +83,12 @@ def cli_args(add=lambda x: None, parser=None, accept_unknown=False):
...
+        default=infra.path.get_default_workspace(),
+    )
+    parser.add_argument("--label", help="Unique identifier for the test", required=True)
```
Comment (achamayou) [inline tests/e2e_args.py]: That should not be different from the perf-label imo.

File: tests/infra/path.py [hunk 6]
```diff
@@ -33,6 +33,10 @@ def build_bin_path(bin_name, enclave_type=""):
         return bin_name
 
 
+def get_default_workspace():
```
Comment (achamayou) [inline tests/infra/path.py]: default_workspace()

File: tests/infra/path.py [hunk 7]
```diff
@@ -33,6 +33,10 @@ def build_bin_path(bin_name, enclave_type=""):
...
+    return os.path.join(os.getcwd(), "workspace")
```
Comment (achamayou) [inline tests/infra/path.py]: default_workspace()

File: tests/infra/remote.py [hunk 8]
```diff
@@ -318,6 +325,10 @@ def start(self):
             stderr=self.stderr,
             env=self.env,
         )
+        if wait_for_node_termination:
```
Comment (achamayou) [inline tests/infra/remote.py]: Best to split that into its own method,

Comment (achamayou) [reply tests/infra/remote.py]: Also this needs timeout logic,

File: tests/infra/remote.py [hunk 9]
```diff
@@ -418,45 +435,60 @@ def __init__(
...
-            if not notify_server_host or not (
-                notify_server_port and notify_server_port[0]
-            ):
+        cmd = [self.BIN, "--enclave-file={}".format(lib_path)]
+
+        # If the remote needs to verify the quote, only a subset of arguments are required
+        if self.verify_quote:
+            cmd += ["--start=verify"]
+
+            if other_quote is None or other_quoted_data is None:
```
Comment (achamayou) [inline tests/infra/remote.py]: if not other_quote or not other_quoted_data is probably enough here, but it's best to split the check to give a more specific exception about which is missing.

File: tests/infra/remote.py [hunk 10]
```diff
@@ -418,45 +435,60 @@ def __init__(
...
+                    "Quote and quoted data should be specified when starting remote in verify mode"
                 )
+            cmd += [
+                "--quote-file={}".format(other_quote),
+                "--quoted-data={}".format(other_quoted_data),
+            ]
+        else:
+            cmd = [
+                self.BIN,
+                "--enclave-file={}".format(lib_path),
```
Comment (achamayou) [inline tests/infra/remote.py]: f"--enclave-file={lib_path}" etc

File: tests/infra/remote.py [hunk 11]
```diff
@@ -551,8 +583,16 @@ def get_ledger(self):
         self.remote.get(self.ledger_file_name)
         return self.ledger_file_name
 
-    def get_ledger_full_path(self):
-        return self.remote.root + "/" + self.ledger_file_name
+    def get_ledger_path(self):
```
Comment (achamayou) [inline tests/infra/remote.py]: ledget_path() etc

File: tests/quote_verification.py [hunk 12]
```diff
@@ -4,125 +4,66 @@
...
+            verifying_node.start(
+                args.package,
+                args.enclave_type,
+                args.workspace,
+                args.label,
+                quote_path,
+                quoted_path,
+            )
+            failed = should_fail
+        except Exception:
```
Comment (achamayou) [inline tests/quote_verification.py]: That seems very vague, we probably have more specific failure expectations?

File: sphinx/source/demo.rst (hunk 2 above)
Comment (jumaffre) [reply sphinx/source/demo.rst]: Oops. I renamed it in python but not here.

**Pull request merged**